### ISSUE TITLE
libopenshot build on Windows: which document is the right one?

### ISSUE THREAD

**Issue by radiolondra** [OpenShot/libopenshot#737]
I'm starting to build libopenshot on Windows 10, but I'm a bit confused about which document to follow to prepare the environment and build the libraries (libopenshot and libopenshot-audio).

This one? https://github.com/OpenShot/libopenshot/wiki/Windows-Build-Instructions
or this?  https://www.openshot.org/static/files/libopenshot/md__home_gitlab-runner_builds_d0022447_0_OpenShot_libopenshot_doc_INSTALL-WINDOWS.html
or this? https://github.com/OpenShot/libopenshot/blob/develop/doc/INSTALL-WINDOWS.md

In any case, all of them are giving issues in a way or another. For example the document from Wiki speaks about how to build Resvg (no other doc is speaking about it), but the result of the build is: 

error: Package `resvg-capi v0.18.0 (C:\msys64\home\Roberto\resvg\c-api)` does not have the feature `qt-backend`

This probably happens because of the version used.... but which version have I to use?

But mostly... which document have I to follow to build the libs?

Could someone power on the light? I'm really confused.
Thanks

**Comment by ferdnyc** [comment]
@radiolondra 

As of just today, the single most accurate and current "documentation" for building libopenshot-audio and libopenshot on Windows is [the Github Actions workflow file](https://github.com/ferdnyc/libopenshot/blob/ci-mac-win/.github/workflows/ci.yml) from my branch for PR #734, which is nearing completion and will be merged in the coming days.

In it, I do just that: build both libopenshot-audio, and then libopenshot, on all three supported platforms, from scratch in an isolated CI environment. (So, there can't possibly be any missing steps or unavailable packages.)

You can check out the [latest Actions runs](https://github.com/OpenShot/libopenshot/runs/3681139698) for Windows to see the build in action, but basically the steps boil down to:
1. Set up MSYS2
2. Clone both repos
3. Install dependencies using `pacman`:
    ```bash
    pacman -S \
              mingw-w64-x86_64-gcc \
              mingw-w64-x86_64-lld \
              mingw-w64-x86_64-make \
              mingw-w64-x86_64-cmake \
              mingw-w64-x86_64-pkgconf \
              mingw-w64-x86_64-qt5 \
              mingw-w64-x86_64-libvpx \
              mingw-w64-x86_64-ffmpeg \
              mingw-w64-x86_64-zeromq \
              mingw-w64-x86_64-opencv \
              mingw-w64-x86_64-protobuf \
              mingw-w64-x86_64-babl \
              mingw-w64-x86_64-catch \
              mingw-w64-x86_64-python3 \
              mingw-w64-x86_64-swig
    ```
    (Babl isn't technically in use yet, but will likely be soon with the merge of #716.)
1. Configure and compile libopenshot-audio
    ```bash
    cd /path/to/libopenshot-audio
    cmake -B build -S . -G "MinGW Makefiles"
    cmake --build build
    ```
1. Configure and compile libopenshot, using the audio build:
    ```bash
    cd /path/to/libopenshot
    cmake -B build -S . -G "MinGW Makefiles" \
      -DOpenShotAudio_ROOT=/path/to/libopenshot-audio/build \
      -DCMAKE_EXE_LINKER_FLAGS=-fuse-ld=lld \
      -DCMAKE_SHARED_LINKER_FLAGS=-fuse-ld=lld
    cmake --build build
    ```

...And you should end up with two DLLs (`build/src/libopenshot.dll` and `build/src/protobuf_messages/libopenshot_protobuf.dll`), a Python extension module (`build/bindings/python/_openshot.pyd`), and a Python wrapper module (`build/bindings/python/openshot.py`). There's also the `build/libopenshot-audio.dll` from the audio library build.

**Comment by ferdnyc** [comment]
(The `-fuse-ld=lld` flags are a very recent addition, as `lld` appears to be **far** faster than the GNU linker on MinGW. Linking all twenty-two unit test runners is a **painfully** glacial process, using GNU ld. My goal — which is still a ways off, sadly — is to hopefully somehow get the Windows CI build runtimes _below_ 30 minutes each. :roll_eyes: )

**Comment by ferdnyc** [comment]
Whoops, you can actually leave `mingw-w64-x86_64-clang` off that list, as libopenshot currently refuses to build with MinGW clang. So you have to use `g++`. (List updated.)

**Comment by radiolondra** [comment]
Thanks @ferdnyc using this way libopenshot-audio successfully compiles...
But... running the demo app I can't hear any sound :(
Something is wrong? Missed some libs? I tried to install ASIO SDK too and set ASIO_SDK_DIR env variable. Cmake see it but the result is still the same: no sound.

**Comment by ferdnyc** [comment]
@radiolondra 

Nope! Nothing's wrong, the issue is that the demo app is compiled as a MinGW console application and thus can't access the Windows Audio APIs. It is possible to build it as a Win32 graphical application, which _will_ let it access Windows Audio... but then it loses its console output! The demo code wasn't written for Windows and never properly worked there, basically. You just have to trust that when compiled into a Win32 executable, the library will work. :grin: 

(I detail in OpenShot/libopenshot-audio#130 some of the issues with the demo code on Windows, and locally in working with the code I've managed to successfully both (a) patch in the necessary bits of COM to let the console app _detect_ the Windows Audio devices (but it still can't play to them), and (b) patch _out_ support for Windows Audio as the default device type, at which point the app then falls back to DirectSound which it _can_ use for playback. But the latter doesn't represent a real-world case anyone actually wants to use (Windows Audio is the correct interface to use, in real-world scenarios), so I haven't bothered to go much farther down that road.

**Comment by radiolondra** [comment]
Thanks @ferdnyc 
I successfully built libopenshot.dll too. Build time: 4 minutes (Ryzen 9 CPU)
Now I'll start my tests. I'll let you know.
Many many thanks!

**Comment by ferdnyc** [comment]
The first, best test you can do is the Python bindings check. In a MinGW shell:
```bash
cd /path/to/libopenshot
cd build/examples
cp ../bindings/python/*openshot* \
../src/libopenshot.dll \
../src/protobuf_messages/libopenshot_protobuf.dll \
/path/to/libopenshot-audio/build/libopenshot-audio.dll \
.
```
Then run python3:
```python3
$ python3
>>> import openshot
>>> openshot.Version
OpenShotVersion('0.2.7-dev')
```
Because all of your other DLL dependencies are on the `$PATH`, everything will hopefully Just Work™.

Getting applications to launch from the Explorer in Windows without crapping out is significantly trickier, though, and involves significantly more copying-around of DLLs. (And usually a `qt.conf` file as well.)

**Comment by ferdnyc** [comment]
With the new dependency features added to CMake 3.21 ([`install(IMPORTED_RUNTIME_ARTIFACTS ...)` & `install(TARGETS ... RUNTIME_DEPENDENCIES ...)`](https://cmake.org/cmake/help/latest/command/install.html#install), and the Windows-specific [`$<TARGET_RUNTIME_DLLS>`](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#genex:TARGET_RUNTIME_DLLS) especially), it should finally be possible to add (cleanly and without undue hoop-jumping) more automation for a lot of that DLL copying at some point in the future. Both during the build process (where needed, like for unit tests), and then later as part of the install process, like when we want to create a self-contained application directory/package.

**Comment by ferdnyc** [comment]
BTW, if you do want to build a version of `openshot-audio-demo.exe` that plays sound from the MinGW console using DirectSound, the last commit on my fork's `testapp-details` branch, ferdnyc/libopenshot-audio@f6dc4ae36e355a7190a0b4159ba6ff58bd56b6ac, has the changes necessary to disable WASAPI support completely. (Just in the demo app.)

**Comment by radiolondra** [comment]
Python bindings check successful! 🥇 

FYI. I printed on paper this whole post as it's the ONLY really working document. I trashed all the other docs.
Thank you @ferdnyc , you are great!
keep in touch

**Comment by radiolondra** [comment]
@ferdnyc 
Let me reopen this post just to inform about my tests progress with libopenshot.
I'm now able to run the "Python binding checks" from a Console Prompt (Windows 10) and to include libopenshot into a pyCharm project which works fine.

The tricks I used:
1 - I can't use `MSC Python`, I had to use `GCC Python` (the same used by Msys64). This because python binaries were generated by GCC and they are not compatible with normal Python (MSC), giving tons of errors like memory corruptions.
2 - I had to copy a subset of `Msys64\mingw64\bin` (python GCC executables and all the dlls) and a subset of `Msys64\mingw64\lib` (Python 3.9 folder only) to the pyCharm project root, together with the 5 files needed to make the test: `_openshot.pyd, openshot.py, libopenshot.dll, libopenshot-audio.dll, openshot-protobuf.dll`

In the console prompt I added `/myproject/path/bin and /myproject/path/lib to the head of the PATH environment variable`. Then running "python" in console, the GCC version of Python will be ran.

In the pyCharm project I used `/myproject/path/bin/python.exe as project interpreter` and added `/myproject/path/bin and /myproject/path/lib folders to the interpreter environment`.

Both tests work like a charm.

If you have any advice for me, I'll be happy to learn from you. 

My final question could be: Which are the really needed DLLs to have in the bin folder? I tried to check with some dependancy softwares, but they are not exhaustive, always some DLL is missed and I had to include all of them for tests to work.

Many thanks
Rob
